lib/std/Thread/Futex.zig

```diff
@@ -183,58 +170,65 @@ const DarwinFutex = struct {
         //
         // ulock_wait() uses 32-bit micro-second timeouts where 0 = INFINITE or no-timeout
         // ulock_wait2() uses 64-bit nano-second timeouts (with the same convention)
+        const supports_ulock_wait2 = builtin.target.os.version_range.semver.min.major >= 11;
+
         var timeout_ns: u64 = 0;
-        if (timeout) |timeout_value| {
-            // This should be checked by the caller.
-            assert(timeout_value != 0);
-            timeout_ns = timeout_value;
+        if (timeout) |delay| {
+            timeout_ns = delay;
         }
-        const addr = @ptrCast(*const anyopaque, ptr);
-        const flags = darwin.UL_COMPARE_AND_WAIT | darwin.ULF_NO_ERRNO;
+
         // If we're using `__ulock_wait` and `timeout` is too big to fit inside a `u32` count of
         // micro-seconds (around 70min), we'll request a shorter timeout. This is fine (users
         // should handle spurious wakeups), but we need to remember that we did so, so that
-        // we don't return `TimedOut` incorrectly. If that happens, we set this variable to
+        // we don't return `Timeout` incorrectly. If that happens, we set this variable to
         // true so that we we know to ignore the ETIMEDOUT result.
         var timeout_overflowed = false;
+
+        const addr = @ptrCast(*const anyopaque, ptr);
+        const flags = os.darwin.UL_COMPARE_AND_WAIT | os.darwin.ULF_NO_ERRNO;
         const status = blk: {
-            if (target.os.version_range.semver.min.major >= 11) {
-                break :blk darwin.__ulock_wait2(flags, addr, expect, timeout_ns, 0);
-            } else {
-                const timeout_us = std.math.cast(u32, timeout_ns / std.time.ns_per_us) catch overflow: {
-                    timeout_overflowed = true;
-                    break :overflow std.math.maxInt(u32);
-                };
-                break :blk darwin.__ulock_wait(flags, addr, expect, timeout_us);
+            if (supports_ulock_wait2) {
+                break :blk os.darwin.__ulock_wait2(flags, addr, expect, timeout_ns, 0);
             }
+
+            const timeout_us = std.math.cast(u32, timeout_ns / std.time.ns_per_us) catch overflow: {
+                timeout_overflowed = true;
+                break :overflow std.math.maxInt(u32);
+            };
+
+            break :blk os.darwin.__ulock_wait(flags, addr, expect, timeout_us);
         };
 
         if (status >= 0) return;
         switch (@intToEnum(std.os.E, -status)) {
+            // Wait was interrupted by the OS or other spurious signalling.
             .INTR => {},
             // Address of the futex is paged out. This is unlikely, but possible in theory, and
             // pthread/libdispatch on darwin bother to handle it. In this case we'll return
             // without waiting, but the caller should retry anyway.
             .FAULT => {},
```

Comment:
stupid question: how does the caller know if INTR or FAULT happened? Again reading `status` sounds wasteful or is it not?
It should be documented at the function doc that there is stuff to handle. Not sure, if also what cases exact.

Reply:
The caller shouldn't need to be aware of it for correctness when using the Futex API given spurious wakes are a possibility. Tis documented at the function level [here](https://github.com/kprotty/zig/blob/323c07084d6a4a2e2ed3eba690eaeba4e6da1c91/lib/std/Thread/Futex.zig#L20).